source/extensions/filters/network/postgres_proxy/postgres_message.h

```diff
@@ -0,0 +1,330 @@
+    pos += sizeof(int32_t);
+    left -= sizeof(int32_t);
+    if (len_ < 1) {
+      // Nothing follows.
+      value_.clear();
+      return true;
+    }
+    if ((left < static_cast<uint64_t>(len_)) ||
+        ((data.length() - pos) < static_cast<uint64_t>(len_))) {
+      return false;
+    }
+    value_.resize(len_);
+    data.copyOut(pos, len_, value_.data());
+    pos += len_;
+    left -= len_;
+    return true;
+  }
+
+  std::string toString() const {
+    char buf[16];
+    sprintf(buf, "[(%d bytes):", len_);
+
+    std::string out = buf;
+    out.reserve(value_.size() * 3 + 16);
+    bool first = true;
+    for (const auto& i : value_) {
+      if (first) {
+        sprintf(buf, "%02d", i);
+        first = false;
+      } else {
+        sprintf(buf, " %02d", i);
+      }
+      absl::StrAppend(&out, buf);
+    }
+    absl::StrAppend(&out, "]");
+    return out;
+  }
+
+private:
+  int32_t len_;
+  std::vector<char> value_;
+};
+
+// Array contains one or more values of the same type.
+template <typename T> class Array {
+public:
+  bool read(const Buffer::Instance& data, uint64_t& pos, uint64_t& left) {
+    // First read the 16 bits value which indicates how many
+    // elements there are in the array.
+    if (((data.length() - pos) < sizeof(uint16_t)) || (left < sizeof(uint16_t))) {
+      return false;
+    }
+    uint16_t num = data.peekBEInt<uint16_t>(pos);
+    pos += sizeof(uint16_t);
+    left -= sizeof(uint16_t);
+    if (num != 0) {
+      for (auto i = 0; i < num; i++) {
+        auto item = std::make_unique<T>();
+        if (!item->read(data, pos, left)) {
+          return false;
+        }
+        value_.push_back(std::move(item));
+      }
+    }
+    return true;
+  }
+
+  std::string toString() const {
+    std::string out;
+    char buf[128];
+    sprintf(buf, "[Array of %zu:{", value_.size());
+    out = buf;
+
+    for (const auto& i : value_) {
+      absl::StrAppend(&out, i->toString());
+    }
+    absl::StrAppend(&out, "}]");
+
+    return out;
+  }
+
+private:
+  std::vector<std::unique_ptr<T>> value_;
+};
+
+// Repeated is a composite type used at the end of the message.
+// It indicates to read the value of the same type until the end
+// of the Postgres message.
+template <typename T> class Repeated {
+public:
+  bool read(const Buffer::Instance& data, uint64_t& pos, uint64_t& left) {
+    if ((data.length() - pos) < left) {
+      return false;
+    }
+    // Read until nothing is left.
+    while (left != 0) {
+      auto item = std::make_unique<T>();
+      if (!item->read(data, pos, left)) {
+        return false;
+      }
+      value_.push_back(std::move(item));
+    }
+    return true;
+  }
+
+  std::string toString() const {
+    std::string out;
+
+    for (const auto& i : value_) {
+      absl::StrAppend(&out, i->toString());
+    }
+    return out;
+  }
+
+private:
+  std::vector<std::unique_ptr<T>> value_;
+};
+
+// Interface to Postgres message class.
+class MessageI {
```

Comment:
Can we just spell this out as `MessageInterface`?

Reply:
Changed it to simply `Message` and marked methods as `PURE`. Most of base classes in Envoy which define API use the same approach.